src/hooks/useDebounceSearch.ts

```diff
@@ -0,0 +1,40 @@
+import { useState, useCallback, useEffect, useMemo } from "react";
+import debounce from "lodash/debounce";
+
+const SEARCH_DEBOUNCE_TIME = 200; // ms
+
+export const useDebounceSearch = () => {
+  const [searchInputValue, setSearchInputValue] = useState("");
+  const [searchQuery, setSearchQuery] = useState("");
+
+  const clearSearchInput = useCallback(() => {
+    setSearchInputValue("");
+    // To instantly remove the options
+    setSearchQuery("");
+  }, []);
+
+  const debouncedSearch = useMemo(
+    () =>
+      debounce((value: string) => {
+        setSearchQuery(value.toLowerCase());
+      }, SEARCH_DEBOUNCE_TIME),
+    []
+  );
+
+  useEffect(() => {
+    debouncedSearch(searchInputValue);
+
+    return () => {
+      debouncedSearch.cancel();
+    };
+  }, [searchInputValue, debouncedSearch]);
```

Comment:
Yes, that's the intended effect. If you don't do it, you would just delay every user's keystroke

Reply:
> If you don't do it, you would just delay every user's keystroke

That doesn't sound right to me..? That's not how debounce tends to work. I've used various debounce functions over the years, and literally never had to use a `.cancel` on them.

Looking at the examples of how to use it, `cancel` is never used, implying it's not standard 'required' usage?

Reply:
> If there is a new keystroke, then that resets the timer

Apparently, this is what I achieved with `.cancel`. Without it, it just postponed every keystroke

Reply:
> Without it, it just postponed every keystroke

That doesn't seem right..? Unless debounce is being run on the thing updating your 'input state' (rather than the 'search state') now that you changed the order of things?

Reply:
> Unless debounce is being run on the thing updating your 'input state' (rather than the 'search state') now that you changed the order of things

There is a useEffect, which is effected by `input state`, but inside runs denounce for `search state
![image](https://user-images.githubusercontent.com/63194656/117118755-d7f9b580-ad99-11eb-9ea9-49b2b9c9f416.png)
`

Actually, it looks like it works without cancel

Reply:
Removed in https://github.com/sparkletown/sparkle/pull/1299/commits/9acbbec8f58d7d7b122f095b27d50c10400627c7

Reply:
Cool, glad it works now!